pairing_api/src/pairing.rs

```diff
@@ -0,0 +1,457 @@
+            sym_key: parsed.sym_key,
+            version: parsed.version,
+            pairing: pairing_info,
+        })
+    }
+}
+
+/// Client for managing WalletConnect pairings.
+#[derive(Debug, Default)]
+pub struct PairingClient {
+    pub pairings: Arc<Mutex<HashMap<String, Pairing>>>,
+}
+
+impl PairingClient {
+    /// initializes the client with persisted storage and a network connection
+    pub fn new() -> Self {
+        Self {
+            pairings: Default::default(),
+        }
+    }
+
+    /// Attempts to generate a new pairing, stores it in the client's pairing
+    /// list, subscribes to the pairing topic, and returns the necessary
+    /// information to establish a connection.
+    pub async fn create(
+        &self,
+        metadata: Metadata,
+        methods: Option<Methods>,
+    ) -> Result<(Topic, String), PairingClientError> {
+        let expiry = Utc::now().timestamp() as u64 + EXPIRY_5_MINS;
+
+        let topic = Topic::generate();
+        let relay = Relay {
+            protocol: RELAY_PROTOCOL.to_owned(),
+            data: None,
+        };
+        let sym_key = gen_sym_key();
+        let pairing_info = PairingInfo {
+            active: false,
+            methods: methods.unwrap_or(Methods(vec![])),
+            expiry,
+            relay,
+            topic: topic.clone().to_string(),
+            peer_metadata: Some(metadata),
+        };
+
+        let uri = Self::generate_uri(&pairing_info, &sym_key);
+        let pairing = Pairing {
+            sym_key,
+            version: VERSION.to_owned(),
+            pairing: pairing_info,
+        };
+
+        {
+            let mut pairings = self.pairings.lock().await;
+            pairings.insert(topic.clone().to_string(), pairing);
+        }
+
+        Ok((topic, uri))
+    }
+
+    /// for responder to pair a pairing created by a proposer
+    pub async fn pair(&self, url: &str, activate: bool) -> Result<Topic, PairingClientError> {
+        let mut pairing = Pairing::try_from_url(url)?;
+        let topic = pairing.pairing.topic.clone();
+
+        {
+            let mut pairings = self.pairings.lock().await;
+
+            // Check if the pairing already exists
+            if let Some(existing_pairing) = pairings.get_mut(&topic) {
+                // Reactivate the pairing if needed
+                if activate {
+                    existing_pairing.pairing.active = true;
+                }
+
+                // If the pairing is already active, return an error
+                if existing_pairing.pairing.active {
+                    return Err(PairingClientError::PairingTopicAlreadyExists);
+                }
+            }
+        }
+
+        {
+            // Activate the pairing if requested
+            if activate {
+                pairing.pairing.active = true;
+            }
+
+            let mut pairings = self.pairings.lock().await;
+            pairings.insert(topic.clone(), pairing);
+        }
+
+        Ok(topic.into())
```

Comment:
There are some issues. 

You first set the pairing as `active=true` if activate param is true. Then, you immediately return an error if it’s already active.

Also you lock pairings twice. This leads to unnecessary complexity and overhead, and there is a small risk that other tasks might acquire the lock between the drop() and subsequent lock(), resulting in race conditions. Its better to simplify the locking by acquiring the lock once and doing all necessary operations before releasing it.

In addition, lets return topic yearly if pairing already exists?

pair function can be fixed like this
```rust
    pub async fn pair(&self, url: &str, activate: bool) -> Result<Topic, PairingClientError> {
        let mut pairing = Pairing::try_from_url(url)?;
        let topic = pairing.pairing.topic.clone();

        let mut pairings = self.pairings.lock().await;

        // Check if the pairing already exists
        if let Some(existing_pairing) = pairings.get_mut(&topic) {
            // If the pairing is already active, return an error
            if existing_pairing.pairing.active {
                return Err(PairingClientError::PairingTopicAlreadyExists);
            }

            // Reactivate the pairing if needed
            if activate {
                existing_pairing.pairing.active = true;
            }

            return Ok(topic.into());
        }

        // Activate the pairing if requested
        if activate {
            pairing.pairing.active = true;
        }

        pairings.insert(topic.clone(), pairing);

        Ok(topic.into())
    
````

Reply:
Q: when we reactivate existing pairing, shouldn't we also update expiring? Like we do in `activate` function

https://github.com/KomodoPlatform/WalletConnectRust/blob/c02d999db2422308ee558172cc852ed62910204a/pairing_api/src/pairing.rs#L219-L230

Reply:
even tho I think they serve two different purposes, I believe we should update expiry too! thanks

Reply:
last note related to this fix.
lets move creating `expiry` logic to standalone priv function, as this code is duplicated 
```rust
        let expiry = Utc::now().timestamp();
        if expiry < 0 {
            return Err(PairingClientError::TimeError(
                "Negative timestamp".to_string(),
            ));
        };
```

Reply:
thanks, resolved